airbyte-config/config-persistence/src/main/java/io/airbyte/config/persistence/StatePersistence.java

```diff
@@ -0,0 +1,280 @@
+/*
+ * Copyright (c) 2022 Airbyte, Inc., all rights reserved.
+ */
+
+package io.airbyte.config.persistence;
+
+import static io.airbyte.db.instance.configs.jooq.generated.Tables.STATE;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import io.airbyte.commons.json.Jsons;
+import io.airbyte.config.StateType;
+import io.airbyte.config.StateWrapper;
+import io.airbyte.db.Database;
+import io.airbyte.db.ExceptionWrappingDatabase;
+import io.airbyte.protocol.models.AirbyteGlobalState;
+import io.airbyte.protocol.models.AirbyteStateMessage;
+import io.airbyte.protocol.models.AirbyteStateMessage.AirbyteStateType;
+import io.airbyte.protocol.models.AirbyteStreamState;
+import io.airbyte.protocol.models.StreamDescriptor;
+import java.io.IOException;
+import java.time.OffsetDateTime;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.UUID;
+import java.util.stream.Collectors;
+import org.jooq.DSLContext;
+import org.jooq.JSONB;
+import org.jooq.Record;
+import org.jooq.RecordMapper;
+import org.jooq.impl.DSL;
+
+public class StatePersistence {
+  private final ExceptionWrappingDatabase database;
+
+  public StatePersistence(final Database database) {
+    this.database = new ExceptionWrappingDatabase(database);
+  }
+
+  /**
+   * Get the current State of a Connection
+   *
+   * @param connectionId
+   * @return
+   * @throws IOException
+   */
+  public Optional<StateWrapper> getCurrentState(final UUID connectionId) throws IOException {
+    final List<StateRecord> records = this.database.query(ctx -> getStateRecords(ctx, connectionId));
+
+    if (records.isEmpty())
+      return Optional.empty();
+
+    return switch (records.stream().findFirst().get().type) {
```

Comment:
Should we have some validatino that ensure that all the records have the same type here?

Reply:
I was wondering about that. I will add a check and log an error.

Not sure what is a desirable behavior here. If we have an inconsistent state in the DB, throwing an error means that recovering from this state will hard since the reset should rely on this same function.
Did we have something in mind to handle those type of errors?

Reply:
I added a check to make it fail harder when we read or write.
Checking before updating the state should prevent us from landing in a bad state.

Reply:
I think this would worth having a team discussion around that because it feels that we are heading to a potential state similar to the quarantine where a connection is blocked and can only be unblocked with a manual intervention. Should we make sure that the write clear any other state if there is a conflict? @cgardens @lmossman @jdpgrailsdev WDYT?

Reply:
@lmossman I think that it might lead to duplication of data in case of the Incremental | Append. And if we introduce an error, all the connection that are running will be fully refreshed which can be very costly. I would prefer to fail when getting and/or saving the state.